modules/infra/control/ctlplane.c

```diff
@@ -0,0 +1,402 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2025 Christophe Fontaine
+
+#include "ctlplane.h"
+
+#include <gr_bond.h>
+#include <gr_config.h>
+#include <gr_control_input.h>
+#include <gr_eth.h>
+#include <gr_event.h>
+#include <gr_iface.h>
+#include <gr_log.h>
+#include <gr_macro.h>
+#include <gr_mempool.h>
+#include <gr_module.h>
+#include <gr_netlink.h>
+#include <gr_nh_control.h>
+#include <gr_port.h>
+#include <gr_vlan.h>
+
+#include <event2/event.h>
+#include <rte_errno.h>
+#include <rte_ethdev.h>
+#include <rte_malloc.h>
+#include <rte_net.h>
+
+#include <errno.h>
+#include <fcntl.h>
+#include <linux/ethtool.h>
+#include <linux/if_tun.h>
+#include <linux/sockios.h>
+#include <net/if.h>
+#include <sys/ioctl.h>
+
+#define TUN_TAP_DEV_PATH "/dev/net/tun"
+
+static struct rte_mempool *cp_pool;
+static struct event_base *ev_base;
+
+static control_input_t port_output_id;
+
+static void finalize_fd(struct event *ev, void * /*priv*/) {
+	int fd = event_get_fd(ev);
+	if (fd >= 0)
+		close(fd);
+}
+
+void iface_cp_tx(struct rte_mbuf *m) {
+	struct mbuf_data *d = mbuf_data(m);
+	struct iface_stats *stats;
+	char *data = NULL;
+
+	if (d->iface->cp_fd == 0)
+		goto end;
+
+	if (rte_pktmbuf_linearize(m) == 0) {
+		data = rte_pktmbuf_mtod(m, char *);
+	} else {
+		data = rte_malloc(NULL, rte_pktmbuf_pkt_len(m), 0);
+		if (data == NULL) {
+			LOG(ERR, "rte_malloc failed %s", rte_strerror(rte_errno));
+			goto end;
+		}
+		// with a non-contiguous mbuf, rte_pktmbuf_read returns a pointer
+		// to the user provided buffer.
+		rte_pktmbuf_read(m, 0, rte_pktmbuf_pkt_len(m), data);
+	}
+
+	// Do not retry even in case of  if EAGAIN || EWOULDBLOCK
+	// If the tun device queue is full, something really bad is
+	// already happening on the management plane side.
+	if (write(d->iface->cp_fd, data, rte_pktmbuf_pkt_len(m)) != rte_pktmbuf_pkt_len(m)) {
+		// The user messed up and removed the tap interface
+		// release resources on our side to try to recover
+		if (errno == EBADF) {
+			iface_destroy(d->iface->id);
+		}
```

Comment:
I just noticed this. Isn't that a bit extreme? Note that it may not be possible to delete the actual interface (for example, removing a port which has vlan interfaces attached to it isn't possible).

Maybe just a loud error message would be enough? Or we could re-create the deleted tap?